Paper: Challenges and Implementations for ML Inference in High-energy Physics (#1102)
scipy-conference/scipy_proceedings: papers/sofie/main.md

---
title: Challenges and Implementations for ML Inference in High-energy Physics

> abstract: |

Reviewer: Consider condensing the abstract. Ideally, aim for a single paragraph. If that’s not feasible, restructuring it into two focused paragraphs would help-- one to introduce the broader problem and another to detail how this paper specifically addresses it (you can also provide a brief about what each section of the paper explores and the overall flow of the paper).

Author: made the change, thanks!

>   At CERN, machine learning models are developed and deployed for a range of applications, including data analysis, event reconstruction, and classification. These models must be not only highly sophisticated but also optimized for efficient inference. One critical application is in trigger systems—designed to identify and select interesting events from an immense stream of experimental data. Experiments like ATLAS and CMS generate data at rates approaching 100 TB/s, requiring triggers to rapidly filter out irrelevant events.

Reviewer: A reference for the "100 TB/s" figure/claim would be useful here.

Author: Probably better to not include references in the abstract, I have added a reference in the HL-LHC section.

>   With the advent of the upcoming High-Luminosity phase of the Large Hadron Collider at CERN, the number of collisions will rise, thereby the data generation rates will further increase significantly. This requires major work to make the architectures for modelling and analysing the involved physical phenomenon further efficient and optimized.

Reviewer: making it brief :
```suggestion
  With the advent of the upcoming High-Luminosity phase of the Large Hadron Collider at CERN, the number of collisions will rise, leading to a significant further increase in data generation rates. This requires major work to further optimize the architectures for modelling and analysing the underlying physical phenomena..
```

>   With the advent of the upcoming High-Luminosity phase of the Large Hadron Collider at CERN, the number of collisions will rise, thereby the data generation rates will further increase significantly. This requires major work to make the architectures for modelling and analysing the involved physical phenomenon further efficient and optimized.

Reviewer 2: Would be helpful to have a reference to any literature on "High-Luminosity phase of the Large Hadron Collider" to provide context to non-experts

Author: added reference

>   To address the requirements of deploying machine learning models in such high-throughput environments, the ML4EP team at CERN has been developing SOFIE—a tool that translates machine learning models into standalone C++ code that is easy to integrate, highly customizable, and optimized for inference. SOFIE is further integrated within the ROOT project and therefore provides a Python API for easier usage within machine learning pipelines. In this work, we review the challenges of running machine learning in production at large-scale experiments like those at CERN and explore solutions such as SOFIE to enhance performance and reliability.

Reviewer: please use the full expansion when the first time an acronym is used-- like `SOFIE—a`, `SOFIE`

Author: made the change, thanks!

---

> ## Introduction

Reviewer: In the introduction, it would be helpful to briefly clarify what types of ML inferences are being referred to in the context of HEP. Are these primarily related to classification, regression, anomaly detection, or other tasks? It might be nice to explain this with an example. You can describe what kind of problem exists in HEP and what softwares exists(that you are talking about below) and what are their limitations and how sofie is a good fix here. Right now introduction seems too vague for someone who is not a domain expert.

Author: Added the details for various ML use-case at the experiments at CERN. Thanks for the suggestion!

> Machine Learning inference is becoming increasingly critical across various domains, particularly in HEP, where efficient model evaluation is essential for production workflows. Integrating inference seamlessly into existing software systems, such as reconstruction, simulation, and analysis software, requires support for evaluating models directly within C++ code,

Reviewer 3: The first use of the acronym - HEP - needs expanded form here.

Author: fixed that, thanks!

beyond the typical Python-based ML environments. Furthermore, effective thread management is crucial for leveraging models in multi-threaded environments, ensuring optimal performance in large-scale data processing tasks. In many HEP applications, inference must be

> performed at the event level, often requiring single-batch processing while maintaining both

Reviewer 3: >> In many HEP applications, inference must be
performed at the event level,  ... << 

--- one or two examples of this would help an uninitiated reader.

Author: added example

computational speed and memory efficiency. Addressing these challenges is key to enabling
fast and resource-efficient inference of ML models within complex scientific workflows.

## Background

> While TensorFlow[@tensorflow2015-whitepaper] and PyTorch[@10.5555/3454287.3455008] provide robust inference capabilities, their use in C++

Reviewer 3: Space between previous word and open paranthesis, (,  for both the references above.

Author: made the change, thanks!

environments presents several challenges. These frameworks are primarily designed around
their native model formats, limiting flexibility when integrating externally trained models.
Using TensorFlow within a C++ environment is particularly challenging, as its C++ API is
not trivial to use and introduces significant dependencies, making deployment more complex.

> Additionally, TensorFlow’s thread management can be challenging to control, and its inference engine is often not optimized for specific use cases, such as single-event evaluation in

Reviewer: - Consider briefly describing what "single-event evaluation" means, or rephrasing it to make the concept clearer. The term may not be immediately understandable to readers unfamiliar with HEP-specific workflows.
```suggestion
Additionally, TensorFlow’s thread management can be challenging to control, and its inference engine is often not optimized for specific use cases, such as that requiring fast, one-at-a-time processing of individual collision events, as is common in HEP workflows.
```

- It might also be interesting to include a brief note(here or in appendix, maybe) or reference explaining how TensorFlow’s thread management presents challenges in these scenarios—for readers curious about the technical limitations, if you think it might be relevant here.

Author: made the change, thanks!

HEP workflows.  

PyTorch, on the other hand, offers the Torch C++ library (LibTorch), which provides a more
convenient interface for C++ integration. It is generally easier to install and requires fewer
dependencies compared to TensorFlow. However, full support for all PyTorch extensions is

> not always available, particularly for specialized libraries such as PyTorch Geometric or PyTorch Cluster, which are commonly used for Graph Neural Networks. Furthermore, certain

Reviewer 2: Could be helpful to provide a more exhaustive list of missing functionalities in existing PyTorch extensions for research tasks to help strengthen the evidence. Would be even better to provide references to published literature using GNN

Author: AFAIK it is not a missing functionality, but more like such extensions have more python heavy APIs therefore not accessible through Libtorch which is a C++ interface.

We only mentioned PyG since it is widely used in the development of GNNs, other extensions are more integrated within the PyTorch ecosystem, and their usage in the Physics experiments varies according to the experiments' usage.

I am not sure what the `references to published literature using GNN` mean, do you imply to provide references to models developed using PyG that are used in High-energy physics research?

> issues arise when converting models from ONNX to the Torch format, limiting the flexibility

Reviewer: - could put an example here of when or why would someone be `converting models from ONNX to the Torch format` in HEP workflows?

- ONNX : full form at first occurrence

Author: made the change, thanks!

of model deployment. These constraints highlight the need for a lightweight and efficient

> inference solution that seamlessly integrates into C++-based data processing pipelines while

Reviewer 3: Weird to read `C++-` Removing the hyphen here would improve readability.

Author: made the change, thanks!

maintaining high performance and flexibility.

> The Open Neural Network Exchange (ONNX)[@bai2019] provides a standardized format for describing and sharing deep learning models, facilitating interoperability across different frameworks. However, ONNX cannot fully represent all model architectures, particularly those

Reviewer 3: Move the expanded form of ONNX at the first use of the acronym. 
Also a white space before each open parenthesis.

Author: made the change, thanks!

used in Graph Neural Networks. To enable the efficient inference of ONNX models, Microsoft developed ONNX Runtime[@onnxruntime], an open-source inference engine that supports both
C++ and Python environments. It offers flexibility by running on both CPUs and GPUs, with

> NVIDIA GPU acceleration via TensorRT[5] and AMD support through ROCm[6].

Reviewer 3: The references [5] and [6] are missing.

Author: corrected that

ONNX Runtime has already been successfully integrated into HEP software frameworks, including ATLAS and CMS, where its convenient C++ API and fine-grained thread control
have proven valuable. As it is based on the ONNX format, trained models from TensorFlow
and PyTorch can be converted for use with ONNX Runtime. However, not all models are fully
compatible with ONNX, posing limitations when working with certain architectures. Despite
these constraints, ONNX and ONNX Runtime provide a promising solution for deploying

> machine learning models efficiently within C++-based scientific computing workflows.

Reviewer 3: Same as above. For readability, replace `C++-` to `C++ `

Author: fixed that, thanks

> ### AthenaTriton

Reviewer: Consider adding transition sentences to smoothly lead into the sections on AthenaTriton and other machine learning inference tools. A brief introductory paragraph summarising the four tools--highlighting their use cases, motivation/contexts, key trade-offs, etc.--would provide helpful context and enhance the flow and structure of the paper. Transitions between tools (e.g., AthenaTriton --> SONIC --> so on) would also improve flow and help the reader understand how these solutions relate or differ.

Author: restructured the overall format, thanks for the suggestion.

> AthenaTriton is a machine learning inference tool developed for the ATLAS experiment to enable scalable inference within the Athena software framework by leveraging the NVIDIA Triton Inference Server. It implements the Inference as a Service IaaS model, where Athena acts as a client that sends requests to a local or remote Triton server for executing ML models. This architecture supports dynamic offloading of computation to GPUs and improves overall resource utilization and load balancing. AthenaTriton uses the AthInferenceTritonTool, which conforms to the common IAthInferenceTool interface, and requires only the model name and server URL to operate. It converts input tensors to raw uint8_t data, sends them via gRPC, and returns the inference outputs in user-defined formats. Triton supports multiple ML backends including ONNX Runtime, TensorRT, TensorFlow, and PyTorch, as well as custom Python and C++ backends, enabling flexible deployment. The tool has been demonstrated using a GNN-based track-finding pipeline called GNN4ITk as a Service, which processes ATLAS ITk simulated data and shows efficient scaling on NVIDIA A100 GPUs. Performance evaluations using perf_analyzer revealed that with increasing concurrent model instances, throughput scaling efficiency remained above 98% while GPU utilization approached 45%. End-to-end tests with Athena clients showed that three concurrent threads achieved a 2.4× speedup, with strong scaling efficiency maintained. AthenaTriton thus provides a robust and maintainable solution for integrating ML inference into both online and offline ATLAS workflows.

Reviewer: - Consider adding some references(e.g., reports, technical notes, or published papers) for the performance claims in this section like "throughput scaling efficiency remained above 98%", "GPU utilization approached 45%", "three concurrent threads achieved a 2.4× speedup".
- Following are some suggestions to improve the readability of the text (you can also consider breaking down this into multiple paragraphs):
```suggestion
AthenaTriton is a machine learning inference tool developed for the ATLAS experiment designed to enable scalable inference within the Athena software framework by leveraging the NVIDIA Triton Inference Server. It follows an Inference-as-a-Service (IaaS) model, where Athena acts as a client that sends requests to a local or remote Triton server for executing ML models. This architecture supports dynamic offloading of computations to GPUs, improving overall resource utilization and load balancing. AthenaTriton uses the AthInferenceTritonTool, which conforms to the common IAthInferenceTool interface.It requires only the model name and server URL to operate. It converts input tensors to raw uint8_t data, sends them via gRPC, and returns the inference outputs in user-defined formats. Triton supports multiple ML backends including, ONNX Runtime, TensorRT, TensorFlow, and PyTorch, as well as custom Python and C++ backends, enabling flexible deployment. The tool has been demonstrated using a GNN-based track-finding pipeline, GNN4ITk-as-a-Service, which processes ATLAS ITk simulated data and shows efficient scaling on NVIDIA A100 GPUs. Performance evaluations using perf_analyzer revealed that, with increasing concurrent model instances, throughput scaling efficiency remained above 98%, while GPU utilization approached 45%. End-to-end tests with Athena clients showed that three concurrent threads achieved a 2.4× speedup, with strong scaling efficiency maintained. AthenaTriton thus provides a robust and maintainable solution for integrating ML inference into both online and offline ATLAS workflows.
```

Reviewer 3: Consider adding decorators like `uint8_t`, `perf_analyzer` for jargon from the code.

> ### SONIC

Reviewer 3: ```suggestion
### Service for Optimized Network Inference on Coprocessors (SONIC)
```

Author: restructured

> The Service for Optimized Network Inference on Coprocessors is a framework adopted by the CMS and ATLAS experiments to facilitate heterogeneous computing through an IaaS. Rather than coupling CPUs directly to coprocessors, SONIC allows for more flexible deployment of ML and non-ML algorithms across specialized hardware such as GPUs, FPGAs, IPUs, and TPUs. This is especially relevant for the HL-LHC phase, where increasing data complexity demands efficient use of computing resources. In CMS, SONIC has been integrated into the MiniAOD data processing workflow using NVIDIA Triton servers to run inference tasks on GPUs, leveraging backends such as ONNX, TensorFlow, PyTorch, and Scikit-learn. Per-model optimization is achieved via Triton’s model analyzer tool, and large-scale tests have been conducted to evaluate performance under simultaneous client loads, with fallback CPU inference servers ensuring robust reliability. Additionally, SONIC supports non-ML acceleration; for example, CMS has implemented Patatrack-AAS and is developing Line Segment Tracking AAS for HL-LHC, both of which enable GPU-based tracking. In ATLAS, SONIC is being developed with the ACTS toolkit for charged particle tracking, including efforts like trac-cc to rewrite ACTS algorithms for GPU execution. The framework involves considerations for backend tuning, batch size optimization, load balancing via Kubernetes, and deployment across multiple sites, all aimed at enhancing scalability, latency performance, and coprocessor saturation in real production environments.

Reviewer: - Consider adding a transition sentence(context) to explain how SONIC connects to the AthenaTriton section-- are they complementary, alternatives, or used in different phases of the workflow? 

- While the paragraph contains valuable technical detail, it may be hard to follow for readers not working with HEP computing. You can simplify or briefly explain terms like:
    - Coprocessors --> “specialized hardware (e.g., GPUs, FPGAs)”
    - Heterogeneous computing --> “using mixed hardware types for better performance”
    - Fallback CPU inference servers  --> “automatically switch to CPUs if GPUs are unavailable”

- Expand acronyms such as MiniAOD, AAS, and HL-LHC on first use.
- Include references/citations for tools and efforts mentioned (e.g., trac-cc, ACTS, Patatrack-AAS, Triton’s model analyzer tool, etc.) to support the technical claims and direct readers for further reading.

Author: restructured this, thanks!

### E2E DL Inference framework in CMSSW

> The End-to-End deep learning inference framework developed within the CMS Software Framework (CMSSW) enables direct use of low-level detector data for machine learning-based object and event classification. Unlike traditional workflows that rely on the particle flow algorithm, which may involve information loss due to simplification, this framework performs inference directly on raw detector inputs, supporting applications such as single-particle classification (electron, photon), jet tagging (quark, gluon, boosted top, tau), and exotic event reconstruction (e.g., H→AA→4γ). The system is structured around three main CMSSW packages—DataFormats, FrameProducer, and Taggers, and operates on detector outputs by generating cropped data frames centered on seed coordinates. These frames serve as inputs to ONNX-converted convolutional neural networks, integrated into the framework using the ONNX C++ API with GPU support. Benchmark studies demonstrate improved latency and throughput when using GPUs compared to CPUs, with event-level inference time reductions of up to 20% for E/Gamma and ~19% for jet-based classifiers. Inference throughput increased by 11–18% across taggers on Fermilab LPC GPUs, with additional scaling tested on NERSC Perlmutter systems. The E2E system is capable of performing stable, reproducible inference with support for future optimization, offering a scalable and high-performance solution for machine learning applications in CMS.

Reviewer: - Consider briefly defining what "low-level detector data" refers to, in 1–2 lines, for clarity.
- Consider adding references for the example applications mentioned-- for interested readers to further read/explore (e.g., jet tagging (quark, gluon, boosted top, tau), and exotic event reconstruction (H→AA→4γ)), the CMSSW packages (DataFormats, FrameProducer, Taggers), and the performance benchmarks.
- The text mentions the use of ONNX-converted convolutional neural networks and integration via the ONNX C++ API, but it’s not clear how this fits into the broader “E2E DL Inference framework.”-- could you please add a line or two clarifying this?

Author: restructured this, thanks!

> The End-to-End deep learning inference framework developed within the CMS Software Framework (CMSSW) enables direct use of low-level detector data for machine learning-based object and event classification. Unlike traditional workflows that rely on the particle flow algorithm, which may involve information loss due to simplification, this framework performs inference directly on raw detector inputs, supporting applications such as single-particle classification (electron, photon), jet tagging (quark, gluon, boosted top, tau), and exotic event reconstruction (e.g., H→AA→4γ). The system is structured around three main CMSSW packages—DataFormats, FrameProducer, and Taggers, and operates on detector outputs by generating cropped data frames centered on seed coordinates. These frames serve as inputs to ONNX-converted convolutional neural networks, integrated into the framework using the ONNX C++ API with GPU support. Benchmark studies demonstrate improved latency and throughput when using GPUs compared to CPUs, with event-level inference time reductions of up to 20% for E/Gamma and ~19% for jet-based classifiers. Inference throughput increased by 11–18% across taggers on Fermilab LPC GPUs, with additional scaling tested on NERSC Perlmutter systems. The E2E system is capable of performing stable, reproducible inference with support for future optimization, offering a scalable and high-performance solution for machine learning applications in CMS.

Reviewer 2: Some references around specialized methods here would be helpful for e.g. particle flow algorithm

Author: added references to the mentioned models and approaches, thanks!

### hls4ml

> hls4ml is an open-source software-hardware codesign workflow designed to bridge the gap between machine learning development and efficient hardware deployment on FPGAs and ASICs. It provides a seamless translation of trained neural networks—developed in standard ML frameworks such as TensorFlow, PyTorch, and QKeras—into HLS implementations suitable for resource-constrained, low-power, or low-latency environments. By supporting optimizations like quantization-aware training, pruning, configurable parallelization, and sparse matrix operations, hls4ml enables the generation of custom hardware accelerators that are tailored to both performance and energy constraints. The framework offers a Python-based API that allows for introspection, bit-accurate emulation, and visualization of the model’s architecture and numerical properties, making it accessible to domain scientists without deep hardware expertise. Its modular structure supports multiple backend tools and vendor-specific workflows, including Xilinx Vivado HLS, Intel Quartus HLS, and ASIC toolchains via Mentor Catapult HLS. As such, hls4ml facilitates rapid prototyping, exploration of design trade-offs, and deployment of machine learning models on edge devices in scientific and embedded computing contexts.

Reviewer: - Consider simplifying or briefly explaining terms like "quantization-aware training" and "bit-accurate emulation" for broader readability.
- Include references to key tools (e.g., hls4ml, etc.)

Author: made the change!

> 

Reviewer: Consider adding a short paragraph-- before going onto the `SOFIE` section-- summarising why existing tools (AthenaTriton, SONIC, E2E, hls4ml) are insufficient for your use case (e.g., C++ integration limits, deployment limits, etc.).

Author: restructured this, thanks!

## SOFIE
To address the challenges of efficient machine learning inference in C++ environments, we

> introduce SOFIE[7]—a tool within ROOT/TMVA[8] designed to generate optimized C++

Reviewer 3: What do [7] and [8] stand for?

Author: corrected that!

code from trained ML models. SOFIE is capable of converting models in ONNX format to
its own Intermediate Representation (Fig. 1). Additionally, it provides limited support for
TensorFlow/Keras and PyTorch models, as well as message-passing Graph Neural Networks
from DeepMind’s Graph Nets library.
The key advantage of SOFIE is its ability to produce standalone C++ code that can be directly invoked within C++ applications with minimal dependencies—requiring only BLAS
for numerical computations. This makes integration seamless for high-energy physics workflows and other computationally demanding applications. Moreover, the generated code can
be compiled at runtime using ROOT[9] Cling Just-In-Time compilation, allowing for flexible
execution, including within Python environments. By eliminating the need for heavyweight
machine learning frameworks during inference, SOFIE offers a highly efficient and easily
deployable solution for ML model evaluation.

### Benchmarking results

> The benchmarking study on SOFIE demonstrates its strong performance in event-level inference tasks common in HEP applications. Conducted on a standard Linux desktop equipped with an AMD Ryzen processor (24 threads, 4.4 GHz) and an NVIDIA RTX 4090 GPU, the benchmarks were run in single-thread mode to ensure a consistent evaluation of inference efficiency across frameworks. SOFIE was tested using both OpenBLAS and Intel MKL for CPU-based linear algebra, and leveraged SYCL with Intel oneAPI and PortBLAS for GPU acceleration. The results show that SOFIE consistently outperforms established frameworks such as ONNX Runtime and LibTorch in scenarios involving small to medium-sized models, particularly for linear architectures and VAEs, achieving lower latency and reduced memory consumption during single-event inference. In these cases, SOFIE benefits from its efficient code generation and use of high-performance mathematical libraries. However, for convolution-heavy models such as Conv2D, Conv3D, and ResNet, SOFIE lags behind due to the absence of optimized convolutional kernels, where LibTorch and ONNX Runtime perform significantly better. For GNNs, SOFIE shows favorable performance on small-scale inputs and maintains competitive scaling behavior, although its efficiency and memory usage degrade with increasing graph size due to the lack of dedicated optimizations. Overall, SOFIE provides a compelling inference solution for C++-based HEP workflows, with minimal dependencies and strong performance on key tasks, and ongoing development efforts aim to further enhance its GPU capabilities, memory optimizations, and support for complex model architectures.

Reviewer: You can consider breaking this paragraph into 3-5 parts to improve readability. Suggested structure:

- First paragraph: machine specifications and benchmarking setup.
- Second para: Presenting benchmarking results.
- Third para: Highlighting where SOFIE performs well 
- Fourth para: Point out where SOFIE underperforms and suggest using other tools for such cases. (points in this paragraph could also be used to transition into the next section on optimisation methods)
- Fifth paragraph(can be included in 4th para): Overview/Summary

Author: restructured the format for the benchmarking results, thanks!

## Optimization methods

> Considering the benchmarking results of SOFIE, we have been working on adding optimization methods in SOFIE, to particularly improve the memory usage of the generated code during inference runtime for very large models. One important aspect of this have been on efficient memory reuse for intermediate tensors needed during the execution. Additionally, fusing machine learning operators further reduces the intermediate memory required by avoiding allocating memory for operations that can be performed in-place. While initially the generated code by SOFIE expanded tensors along dimensions for operations that had compatible but unequal shapes, efficient broadcasting mechanisms were developed to reduce further memory copy. 

Reviewer: ```suggestion
Considering the benchmarking results of SOFIE, we have been working on adding optimization methods in SOFIE, to particularly improve the memory usage of the generated code during inference runtime for large models. One important aspect of this has been on efficient memory reuse for intermediate tensors needed during the execution. Additionally, fusing machine learning operators further reduces the intermediate memory required by avoiding allocating memory for operations that can be performed in-place. While initially the generated code by SOFIE expanded tensors along dimensions for operations that had compatible but unequal shapes, efficient broadcasting mechanisms were developed to reduce further memory copy. 
```

### Memory reuse

> In addition to weights, the execution of machine learning models require several intermediate tensors that handle and carry the results of the intermediate steps of a model execution. Therefore, for a model containing 2 set of layers of Dense and ReLU operations, we will need intermediate memory that shall collect the output from those layers and is supplied as inputs to the next, so a total of 3 intermediate tensors are required.

Reviewer: ```suggestion
In addition to weights, the execution of machine learning models requires several intermediate tensors that handle and carry the results of the intermediate steps of a model execution. Therefore, for a model containing two sets of layers of Dense and ReLU operations, we will need intermediate memory that shall collect the output from those layers and is supplied as inputs to the next, so a total of 3 intermediate tensors are required.
```

Reviewer 3: Contemporary language: 
... we will need intermediate memory that collects the output ... 

instead of shall collect

Author: Hi, we have decided to proceed with this as a survey paper because of the benefits this gives to introduce the various ML Inference approaches to the general scientific community. Thereby, avoiding the detailed information in this paper.

:::{figure} image_memory_reuse_1.png
:label: fig:memory_reuse_1
A model containing 2 sets of Dense and ReLU layers will need a total of 3 intermediate tensors
:::

> Without any optimizations, SOFIE used to allocate memory for each of them, which is unnecessary if some of those tensors will not be required anymore if the execution flow has already progressed. In the example below, since Intermediate Tensor 1 will no longer be required when the execution flow is with the last Gemm Operator, therefore the Intermediate Tensor 3 can safely take its place in the memory instead of allocating additional memory.

Reviewer: improving clarity, conciseness, and readability

```suggestion
Without any optimizations, SOFIE would allocate memory for each intermediate tensor, even when some were no longer needed as the execution progressed. In the example below, since Intermediate Tensor 1 is no longer required once the execution reaches the final GEMM operator, Intermediate Tensor 3 can safely reuse its memory space instead of allocating additional memory.
```

Author: removed this, since the paper shall focus more on discussing the various approaches currently used in High-energy physics.

:::{figure} image_memory_reuse_2.png
:label: fig:memory_reuse_2
In this example, we assume the intermediate tensors are of Float data type and are of length 2 each.
:::

Therefore, we developed a memory reuse mechanism that evaluates memory required and possible positioning ahead of time, which is then used to allocate a big block of memory that can be used to develop pointers to assign memory blocks for the intermediate tensors.
:::{prf:algorithm} Memory reuse mechanism
:label: Memory reuse mechanism

**Inputs** Given a list of operators with input and output tensors, a Total Stack to track the total memory occupied, an Available Stack to track the vacant positions in Total Stack

> **Output** Compute positions to assign for intermediate tensors.

Reviewer: It would be helpful to use variable names that follow common programming conventions to reduce ambiguity-- such as `Total` or `total` or `total_stack`:
```suggestion
**Inputs**: Given a list of operators with input and output tensors, a `Total` stack to track the total memory occupied, and an `Available` stack to track the vacant positions in `Total` stack

**Output**: Compute positions to assign for intermediate tensors.
```

1. For each operator:
  1. For every output tensor:  
    1. Check if Available Stack has any suitable memory chunk  
    2. If yes, reposition it for reuse  
    3. If no, obtain new memory from pool and track it in Total Stack  
  
  2. For every input tensor:  
    1. Check if it is the last operator which is using this as input  
    2. If yes, consider it in available memory (for memory reuse)  
      1. Check if the newly available chunk can be coalesced with an adjoining chunk to make a larger block
:::

### Multi-Operator Fusion

> Several operators in a machine learning model involve weight-less in-place operations, i.e. operatations are element-wise are not dependent on the overall structure of the input tensor, additionally the shapes of their input and output tensors remain unchanged. Such operations include ReLU, BatchNormalization, LayerNormalization, etc., can be fused together so as to avoid allocating memory for their outputs, instead computations can be performed directly on their input tensors which are then dispatched as output for the next operator's utilization.

Reviewer: you can consider breaking this sentence into multiple sentences for clarity. In the suggestion I've included ";" but please feel free to break acc. to what seems apt. to you.
```suggestion
Several operators in a machine learning model involve weight-less in-place operations, i.e., operatations are element-wise and are not dependent on the overall structure of the input tensor; additionally, the shapes of their input and output tensors remain unchanged. Such operations, include ReLU, BatchNormalization, LayerNormalization, etc., can be fused together so as to avoid allocating memory for their outputs; instead computations can be performed directly on their input tensors, which are then dispatched as output for the next operator's utilization.
```

```{admonition} Operator Fusion Algorithm
:class: tip

> The following algorithm describes the logic for operator fusion in a computational graph.

Reviewer: it would be helpful to define the terms like "computational graph"(what are is the set of nodes, edges; properties like DAG, etc.), "anchor operation" before using them in the algorithm.

Author: removed this, to focus on the survey.

1. **For each operator** in the computation graph:
   1. **Check if it is an anchor operation** (e.g., `GEMM`, `Conv`, etc.):
      1. If **yes**:
         1. **Check if the next operator is fusable**, i.e., an in-place, weight-less operation:
            1. If **yes**:
               1. Fuse it with the preceding operation.
               2. **Check if this is the last fusable operation** in the chain:
                  1. If **yes**:
                     1. Break the fusion chain and resume the mechanism from the next operator.
                  2. If **no**:
                     1. Continue to the next operator.
            2. If fusion is **not possible**:
               1. Move to the next operator.
      2. If it is **not an anchor operation**:
         1. Fusion is not possible, move to the next operator.
```
### Efficient Broadcasting

> Operations in the execution of a machine learning model require tensor computations that may involve binary operations, where if the input tensors are of not the same shape, they are broadcasted to a matching shape before being computed. Broadcasting is a process of of extending towards a dimension in a tensor without actually being expanding through copying. While initially the generated code by SOFIE used to expand through un-equal dimensions, a more efficient broadcasting algorithm was developed to avoid further memory allocations and movements.

Reviewer: consider including a reference to the work on "a more efficient broadcasting algorithm" that you mentioned towards the end of the paragraph

```suggestion
Operations in the execution of a machine learning model require tensor computations that may involve binary operations, where if the input tensors are not of the same shape, they are broadcasted to a matching shape before being computed. Broadcasting is a process of extending towards a dimension in a tensor without actually expanding through copying. While initially the generated code by SOFIE used to expand through unequal dimensions, a more efficient broadcasting algorithm was developed to avoid further memory allocations and movements.
```

## Inference on heterogeneous architectures

> We have been working on adding the support in SOFIE for inference on heterogeneous architectures by abstracting device-specific operations through a flexible buffer-accessor model using libraries like SYCL and ALPAKA. At the core of its design is an Infer function that allows users to explicitly control execution on Intel, NVIDIA, or AMD GPUs. During session instantiation, SOFIE initializes input and output buffers, enabling seamless buffer-based data exchange. The backend automatically selects appropriate BLAS routines based on the target execution architecture, ensuring optimal performance without manual intervention. Support for inference on heterogeneous architectures is further enhanced by the use of abstraction libraries such as SYCL and ALPAKA, which offer platform-agnostic interfaces for memory management, making them well-suited for inference code generation in SOFIE. Development is ongoing, with prototypes for SYCL and ALPAKA integration already underway. A CUDA-based prototype supporting GEMM and ReLU operations is currently under testing, while support for Intel and AMD GPUs is planned.

Reviewer: - The paragraph currently mentions SYCL and ALPAKA twice, which seems redundant. Consider merging the two references to these abstraction libraries into a single, focused sentence--preferably early in the paragraph--to avoid repetition and improve flow.

- consider adding references to buffer-accessor model, SYCL, and ALPAKA, and the prototypes if they are open-sourced.

Author: removed this to focus on survey

:::{figure} image_heterogeneous.png
:label: fig:heterogeneous
SOFIE supporting inference of machine learning models on heterogeneous architectures through abstract libraries.
:::

## Experimental Evaluation

> Experimentations were performed on similar initial benchmarking conditions to evaluate the optimization methods developed on the ParticleNet model developed and utilized by the CMS Experiment of CERN. We observed an approximate 50% reduction in the memory usage from SOFIE's previous iteration, but it was observed that SOFIE still requires more memory for its operation in comparison to ONNXRuntime, although SOFIE's inference was time was improved lowering its latency lower than that of ONNXRuntime.

Reviewer: - I've changed "similar" --> "same" ; please reverse that if it's not true, and detail what the differences in the benchmarking conditions are.
```suggestion
Experimentations were performed on same initial benchmarking conditions to evaluate the optimization methods developed on the ParticleNet model developed and utilized by the CMS Experiment of CERN. We observed an approximate 50% reduction in the memory usage from SOFIE's previous iteration, but it was observed that SOFIE still requires more memory for its operation in comparison to ONNXRuntime, although SOFIE's inference time was improved, lowering its latency lower than that of ONNXRuntime.
```

Author: made the change, thanks!

:::{figure} image_eval_optimization.png
:label: fig:evaluation_optim

> Evaluations shows SOFIE's improved memory usage but still requires more memory than ONNXRuntime

Reviewer: ```suggestion
Evaluations show SOFIE's improved memory usage but still require more memory than ONNXRuntime.
```

Author: made the change, thanks!
